lib/src/css_mirroring/bidi_css_gen.dart

```diff
@@ -0,0 +1,355 @@
+        TreeNode usedDirective = mode == RetensionMode.keepFlippedBidiSpecific ? flippedTopLevel : originalTopLevel;
+        _editDirectives(removals, usedDirective, originalTopLevel.rulesets, flippedTopLevel.rulesets, mode, targetDirection);
+      }
+      else if(originalTopLevel.runtimeType == flippedTopLevel.runtimeType && _isDirectionIndependent(originalTopLevel)) {
+        if (mode != RetensionMode.keepBidiNeutral) {
+          _removeRuleSet(removals, iTopLevel);
+        }
+      }
+      else {
+        checkState(originalTopLevel.runtimeType == flippedTopLevel.runtimeType);
+        // TODO: throw sass file error
+        // TODO: handle pagedirective and key frame animation.
+      }
+    }
+  }
+
+  /// Edit the topLevel Ruleset.
+  /// It takes transaction, the topLevels of original and flipped css, Retantion mode, Direction of output css, the index of current topLevel and end of parent of topLevel.
+   _editRuleSet(_PendingRemovals removals, RetensionMode mode, Direction targetDirection, int iTopLevel, int parentEnd) {
+    _storeRemovableDeclarations(removals, _originalTopLevels, _flippedTopLevels, mode, iTopLevel);
+    if (_isRuleRemovable(removals, _originalTopLevels[iTopLevel])) {
+      _removeRuleSet(removals, iTopLevel);
+    }
+    else {
+      removals.commitDeclarations();
+      /// Add direction attribute to RuleId for direction-specific RuleSet.
+      if (mode != RetensionMode.keepBidiNeutral) {
+        _appendDirectionToRuleSet(removals, removals._topLevels[iTopLevel], targetDirection);
+      }
+    }
+  }
+
+  _editDirectives(_PendingRemovals removals, var usedDirective, List<RuleSet> originalRuleSets, List<RuleSet> flippedRuleSets, RetensionMode mode, Direction targetDirection) {
+    var _ruleStartPoints = <int>[];
+    var _ruleEndPoints = <int>[];
+    for(int iRuleSet = 0; iRuleSet < originalRuleSets.length; iRuleSet++) {
+        _storeRemovableDeclarations(removals, originalRuleSets, flippedRuleSets, mode, iRuleSet);
+        if(_isRuleRemovable(removals, originalRuleSets[iRuleSet])) {
+          var startOffset = _getRuleSetStart(originalRuleSets[iRuleSet]);
+          var endOffset = _getRuleSetEnd(usedDirective.rulesets, iRuleSet, usedDirective.span.end.offset);
+          _ruleStartPoints.add(startOffset);
+          _ruleEndPoints.add(endOffset);
+        }
+      else {
+          removals.commitDeclarations();
+          if (mode != RetensionMode.keepBidiNeutral) {
+            _appendDirectionToRuleSet(removals, usedDirective.rulesets[iRuleSet], targetDirection);
+          }
+        }
+    }
+    if(_ruleStartPoints.length == originalRuleSets.length ) { // All rules are to be deleted
+      _removeDirective(removals, usedDirective);
+    }
+    else {
+      _removeStoredRules(removals, _ruleStartPoints, _ruleEndPoints);
+    }
+  }
+
+  _storeRemovableDeclarations(_PendingRemovals removals, var originalTopLevels, var flippedTopLevels, RetensionMode mode, int iTopLevel) {
```

Comment:
comment + naming? I have no clue what this does :-)
(a good comment would document the parameters)

(is this the method that keeps / prunes? store seems the wrong word)